### ISSUE TITLE
Memory Leaks in Native Code

### ISSUE THREAD

**Issue by ironsalsa** [mitmproxy/mitmproxy#3191]
##### Steps to reproduce the problem:

1. Download and extract Linux binaries from https://github.com/mitmproxy/mitmproxy/releases/download/v4.0.1/mitmproxy-4.0.1-linux.tar.gz
2. sudo ./mitmweb --web-iface 192.168.86.88 --web-port 8081 --showhost --listen-host 192.168.86.88 --listen-port 8080
3. Generate load from an external device or from a local browser. I use MITMProxy to capture video traffic, which tears through memory rather quickly against an HLS stream like http://bitdash-a.akamaihd.net/content/sintel/hls/playlist.m3u8. You can install https://addons.mozilla.org/en-US/firefox/addon/native_hls_playback/ in Firefox to get the playback to happen natively.
4. Watch the stream tear through your RAM with .ts fragments (this is expected - video is heavy).
5. Clear the screen with MITMProxy --> New
6. Watch the RAM usage - it won't go down, and will continue to rise when new fragments come in.

##### Any other comments? What have you tried so far?

- Occurs in Docker and running natively tested on both x86 and ARMv7
- Reproduced in Ubuntu 18.04 LTS and Alpine 3.7 Linux. 

I'm wondering if there's a part of the Python API for MITMProxy that MITMWeb needs to add to the "New" instruction in the React code.

##### System information
Mitmproxy: 4.0.1 binary
Python:    3.6.3
OpenSSL:   OpenSSL 1.1.0h  27 Mar 2018
Platform:  Linux-4.15.0-22-generic-x86_64-with-debian-buster-sid

**Comment by madt1m** [comment]
Hey! Thanks for the recap. I have yet to reproduce this by myself, so I will give you more information later.

**Comment by madt1m** [comment]
Ehi! Thanks for the recap. I still have to reproduce this in my machine, so I will have more information later, but File -> New _should_ clear all the flows in memory, so if this does not happen, it could be due to a bug.

> I'm wondering if there's a part of the Python API for MITMProxy that MITMWeb needs to add to the "New" instruction in the React code.

https://github.com/mitmproxy/mitmproxy/blob/09ff5df2fb354a017afa02138b329e615a80c1d0/mitmproxy/addons/view.py#L276

This is the piece of contact that should be correctly called, I believe.

**Comment by mhils** [comment]
@ironsalsa, I'm not very familiar with HLS - is this a single persistent connection, or are we talking about a large number of HTTP requests here?

**Comment by ironsalsa** [comment]
HLS causes a large number of HTTP requests - it progressively requests and loads video fragments while streaming, and the fragments are fairly large themselves.

**Comment by ironsalsa** [comment]
TL;DR What exactly does this line do? I'm not seeing data.clear() as pointing to anything in this scope.
https://github.com/mitmproxy/mitmproxy/blob/e74e97bcd96346a1ae8c372a94149100a755739a/mitmproxy/addons/eventstore.py#L29

Long explanation:
I dug in a few hours understanding the source code flow, and I've either run into a bit I can't understand or possibly the bug itself. Let's follow a clear command start to finish:

* MITMWeb frontend sends the ./clear signal after logic picking the endpoint and adding a token
https://github.com/mitmproxy/mitmproxy/blob/e74e97bcd96346a1ae8c372a94149100a755739a/web/src/js/utils.js#L97-L101
* MITM API is exposing clear through this init
https://github.com/mitmproxy/mitmproxy/blob/e74e97bcd96346a1ae8c372a94149100a755739a/mitmproxy/tools/web/app.py#L512
* That API endpoint triggers ClearAll() without any RequestHandler passed
https://github.com/mitmproxy/mitmproxy/blob/e74e97bcd96346a1ae8c372a94149100a755739a/mitmproxy/tools/web/app.py#L242-L245
* self.view.clear() clears the view
* self.master.events forwards to master.py
* WebMaster().events.clear() is called, passing to EventStore().clear()
https://github.com/mitmproxy/mitmproxy/blob/e74e97bcd96346a1ae8c372a94149100a755739a/mitmproxy/tools/web/master.py#L29
*EventStore().clear() is called
https://github.com/mitmproxy/mitmproxy/blob/e74e97bcd96346a1ae8c372a94149100a755739a/mitmproxy/addons/eventstore.py#L25-L30
*clear() calls self.data.clear(), which does...what? Data.clear() isn't defined here in a way I can see.
https://github.com/mitmproxy/mitmproxy/blob/e74e97bcd96346a1ae8c372a94149100a755739a/mitmproxy/addons/eventstore.py#L29

**Comment by madt1m** [comment]
`self.data` there is a `collections.deque`.

Take a look at:
https://docs.python.org/3.7/library/collections.html#collections.deque

There you can find the specs of `clear()` method. 

Plus, if `view.clear()` is correctly called, the storage layer should be completely wiped, and that's weird given your situation. Did anyone else reproduced this?

**Comment by mhils** [comment]
My suspicion is that we still hold a reference to the flows somewhere in memory. 

One approach to look into this would be to write a mitmproxy add-on that saves a reference to the first flow and then later prints out who else is holding one to the flow after it's been cleared. This can be done with https://docs.python.org/3/library/gc.html#gc.get_referrers.

**Comment by ironsalsa** [comment]
@madt1m Ah, gotcha - didn't recognize clear() as a built in method.
@mhils I'll see what I can do on that front next time I get hours to dig in - I'm out of hours for now.

**Comment by ironsalsa** [comment]
@mhils got a few hours to give the gc.get_referrers script a shot - I'll let y'all know what I get.

**Comment by sloth2012** [comment]
I got the same issue on mitmweb. call the ```view.clear()``` just decrease the speed of memory raises. It will be crashed eventually

**Comment by ironsalsa** [comment]
I have time to try to diagnose again - I'll check back in if I can nail it down.

**Comment by ironsalsa** [comment]
I've created an addon to show what's holding the memory from a flow in MITMWeb - here's the AddOn:

```python
"""
Memory Finder

- Use case: Find out what's holding on to a flow to troubleshoot garbage collection
- Holds a reference to the first flow in a capture
- Prints out who else is holding on to the flow every 5 seconds
"""
import collections
import typing
import time
import gc

import mitmproxy
import mitmproxy.flow

from mitmproxy import connections  # noqa
from mitmproxy import ctx
from mitmproxy import flowfilter
from mitmproxy import exceptions
from mitmproxy import command
from mitmproxy import connections
from mitmproxy.script import concurrent
from mitmproxy import ctx
from mitmproxy import io
from mitmproxy import http  # noqa

class MemFinder:
    def __init__(self):
        self.before_flow = True

    @concurrent
    def request(self, flow):
        """
        Called when a server response has been received.
        """
        # ctx.log.info("Referrer: %d" % self.before_flow)
        # self.before_flow = False
        if self.before_flow:
            self.before_flow = False
            referrers = gc.get_referrers(flow)
            """
            Print out all referrers from a flow every 5 seconds
            """
            while True:
                for referrer in referrers:
#                    ctx.log.info("Referrer: %d" % referrer)
                    print(referrer)
                    time.sleep(5)

addons = [
    MemFinder()
]
```
And here's the referrers I have after using MITMWeb to clear flows after loading http://www.hamsterrepublic.com in Firefox (pointing to the MITMProxy IP and port):
```shell
<frame object at 0x1268f18>
{'obj': <HTTPFlow
  request = Request(GET hamsterrepublic.com:80/)
  client_conn = <ClientConnection: 127.0.0.1:56552>
  server_conn = <ServerConnection: <no address>>>, 'q': <queue.Queue object at 0x7fe85431fbe0>, '_state': 'taken', 'value': <object object at 0x7fe85edd43f0>}
<frame object at 0x7fe853abc930>
(<HTTPFlow
  request = Request(GET hamsterrepublic.com:80/)
  client_conn = <ClientConnection: 127.0.0.1:56552>
  server_conn = <ServerConnection: <no address>>>,)
(<mitmproxy.addons.script.ScriptLoader object at 0x7fe8543529b0>, 'request', <HTTPFlow
  request = Request(GET hamsterrepublic.com:80/)
  client_conn = <ClientConnection: 127.0.0.1:56552>
  server_conn = <ServerConnection: <no address>>>)
^C(<HTTPFlow
  request = Request(GET hamsterrepublic.com:80/)
  client_conn = <ClientConnection: 127.0.0.1:56552>
  server_conn = <ServerConnection: <no address>>>,)
(<__mitmproxy_script__.memory_finder.MemFinder object at 0x7fe854316630>, <HTTPFlow
  request = Request(GET hamsterrepublic.com:80/)
  client_conn = <ClientConnection: 127.0.0.1:56552>
  server_conn = <ServerConnection: <no address>>>)
<cell at 0x7fe853ab3318: HTTPFlow object at 0x7fe85431f7b8>
```

**Comment by ironsalsa** [comment]
From the above, it looks like only 2 things are holding on to the reference:
- The script itself (of course)
- HTTPFlow
Does this help with the troubleshooting, or am I missing something?

**Comment by ironsalsa** [comment]
Updated results comment above with updated logic - now I believe it's accurate to the issue.

**Comment by mhils** [comment]
Thanks, this is very helpful. Looking at the [reference diff](https://gist.github.com/mhils/0ff5abc228741e822a1da0c11c86e20a/revisions#diff-632a2406bbcbcd553eec45ac14b40a0a), we clearly do drop some references, but not all of them.

We hold three references [after clearing the flows](https://gist.github.com/mhils/0ff5abc228741e822a1da0c11c86e20a/a8b5df7d6827aef29e3ac002334876d4d49cf91e
):

```
<frame object at 0x04D52310>
```
This should be memfinder.py itself.

```
[<HTTPFlow
  request = Request(GET example.com:80/)
  response = Response(200 OK, text/html; charset=UTF-8, 1.24k)
  client_conn = <ClientConnection: 127.0.0.1:58275>
  server_conn = <ServerConnection: example.com:80>>]
```

The flow itself (?). Not sure why this is here.

```
{'_state': 'committed',
 'obj': <HTTPFlow
  request = Request(GET example.com:80/)
  response = Response(200 OK, text/html; charset=UTF-8, 1.24k)
  client_conn = <ClientConnection: 127.0.0.1:58275>
  server_conn = <ServerConnection: example.com:80>>,
 'q': <queue.Queue object at 0x04D76610>,
 'value': <HTTPFlow
  request = Request(GET example.com:80/)
  response = Response(200 OK, text/html; charset=UTF-8, 1.24k)
  client_conn = <ClientConnection: 127.0.0.1:58275>
  server_conn = <ServerConnection: example.com:80>>}
```

This is a Reply object. This really shouldn't be around, no idea why it is? We probably need to trace down where and why this is here.



For reference, here's my slightly modified script: https://gist.github.com/mhils/0ff5abc228741e822a1da0c11c86e20a

**Comment by mhils** [comment]
We fooled ourselves here - if we reassign `referrer` and `referrers` we end up with zero references and a cleanup. 


Digging further, here's a script that just counts the total number of tracked objects: https://gist.github.com/mhils/703684f94d6da0b8de11fbee08c07e47
If I run this, then send 20-30 requests through the proxy, and then clear the view, I end up with negligible deltas that don't explain any changes in memory consumption: https://gist.github.com/mhils/ece2c692c81081b9165fd608142562bf/revisions

Long story short, the problem seems to be somewhere in native code. Not sure if there's an efficient way to debug this, I won't pursue this any further for now - sorry.

**Comment by sloth2012** [comment]
@mhils I think may be you are right. I have tried the dumpMaster, it has the same issue. seems that memory increases due to somewhere  in native code

**Comment by mhils** [comment]
This here might be a possible approach forward: https://benbernardblog.com/tracking-down-a-freaky-python-memory-leak/#usingumdhtosearchforcbasedmemoryleaks. It's likely pyOpenSSL or CPython, there's not much other native code in mitmproxy

**Comment by bleuf1sh** [comment]
I've also noticed a memory leak in mitmdump
@mhils , I'm curious how you've been able to run the proxy for more than 90 days.

I concur that it must be somewhere in the native code based off my findings. I made this GIST 
https://gist.github.com/bleuf1sh/f4a4f026ed988fac78183d05342b33f1
where I even remove all the other addons (in case they were the culprit)

I also don't think it is pyOpenSSL because in my tests I was simply refreshing an HTTP url
http://www.un-documents.net/aconf48-14r1.pdf

It leaked at the same rate even when returning a response from within the request hook.

I've also noticed that the memory usage appears to flatline around 256 MB

**Comment by bleuf1sh** [comment]
is there anything we can do about this? I'm happy to help and I've attempted to debug it but I can't seem to find where all the memory is being used up in. 

I need a proxy to help with some testing I'm doing and I'd love to be able to use this project as a solid core. There really isn't anything in the same league.

**Comment by guoyuanjing2988** [comment]
Observed memory leak as well:
Python 3.7.3
mitmproxy: 4.0.4

**Comment by arnulfojr** [comment]
I'm having this issue as well, memory increases real fast!
Python 3.7.3
Mitmproxy 4.0.4 OpenSSL 1.0.2g

Is there a fix any soon? :)

**Comment by nesttle** [comment]
Facing this issue since version 0.18

**Comment by nesttle** [comment]
How can we help?

**Comment by iTosun** [comment]
Could this be the same issue?
https://bugs.python.org/issue34745

**Comment by mhils** [comment]
@iTosun: Unlikely as we don't use asyncio's SSL implementation in mitmproxy.

**Comment by mhils** [comment]
A quick recap of the current state:
  - **mitmproxy** will currently keep all flows in memory unless they are manually cleared from the view. An increase in memory is about to be expected.
  - **mitmdump** (mitmproxy's non-interactive version) should automatically discard flows once they are processed. There should not be any memory leaks.

That being said, we seem to have some memory leak(s) also in mitmdump, which are hard to trace down as they seem to appear in native code (and not Python code). 

**Going ahead**
The good news on that front is that I'm actively rewriting our core proxying code to completely separate I/O and protocol logic (https://github.com/mitmproxy/mitmproxy/projects/4). This is a massive undertaking, but it either (1) eliminates the leak(s) in the process or (2) provide us with a much more fine-grained control over I/O to determine the culprit. If someone wants to join on that effort please feel free to ping me on Slack. It's super challenging, yet super fun!

**Comment by iTosun** [comment]
> @iTosun: Unlikely as we don't use asyncio's SSL implementation in mitmproxy.

Well, if you read the issue you could see that the combination of asyncio and Openssl causes the problem. From what i can read in mitmproxy sourcecode Openssl library is used, so this still could be the same issue

**Comment by guysoft** [comment]
Is there a workaround for this issue? @iTosun notes spesific versions of SSL. Is there a documented version that works?

**Comment by iTosun** [comment]
There should be because the bug is marked as fixed:
https://bugs.python.org/issue34745

Here the commit's that fix the bug from the cpython repo:
- https://github.com/python/cpython/commit/f683f464259715d620777d7ed568e701337a703a
- https://github.com/python/cpython/commit/7f7485c0605fe979e39c58b688f2bb6a4f43e65e

**Comment by iTosun** [comment]
i can also verify that i've successfully ran mitmweb for several hours with consistent memory use and a lot of traffic

**Comment by guysoft** [comment]
I've been using ``3.7.5`` and:
```
mitmdump --version
Mitmproxy: 5.1.1
Python:    3.7.5
OpenSSL:   OpenSSL 1.1.1g  21 Apr 2020
```
And I am getting a large memory leak. 

The only addon I have is this:

```python
from mitmproxy import http

def request(flow: http.HTTPFlow):
    # answer from proxy
    if "&" in flow.request.path:
        print("Redirecting & to correct file")
        flow.request.path = flow.request.path.replace("&", "%26")
```

In less than an hour and heavy traffic of downloading large files I get  a memory leak of a few good gigs.

**Comment by ranjith19** [comment]
any fixes for this?

**Comment by Prinzhorn** [comment]
can you still reproduce this with the current master (snapshots here https://mitmproxy.org/downloads/#branches/master/)? It contains the rewrite of the proxy core and might have solved this. At least it should be easier to track down now.

**Comment by mhils** [comment]
I will close this here for now - as @Prinzhorn mentioned, we have a completely new proxy core. If you still run into memory leaks in practice, take a look at https://github.com/mitmproxy/mitmproxy/issues/4368 to see what an actionable report looks like. 😃
